File: QMOLEDEV/lesstif-0.95.2/lib/Mrm-2.1/misc.c
/**
 *
 * $Id: misc.c,v 1.1 2004/08/28 19:22:35 dannybackx Exp $
 *
 * Copyright (C) 1995 Free Software Foundation, Inc.
 * Copyright (C) 1995-2001 LessTif Development Team
 *
 * This file is part of the GNU LessTif Library.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *
 *  Original author:  Geoffrey W. Ritchey
 *                    codesmit@southwind.net
 *
*/

#include <LTconfig.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>

#include "misc.h"

#include <XmI/DebugUtil.h>


static const int BUFFER_SIZE = 256;
int __MrmStackPointer = 0;
char *__MrmStack[MAX_STACK];


char *
__MrmStore(char *s)
{
    int len = strlen(s);
    static char *buffer = NULL;
    static int SpaceAvailable = 0;
    char *ReturnValue;

    if (len >= SpaceAvailable)
    {
	int BufferSize = BUFFER_SIZE;
	if (len > BUFFER_SIZE)
	    BufferSize = len + 1;
	buffer = (char *)malloc(BufferSize * sizeof(char));
	if (NULL == buffer)
	{
	    __MrmExit(LOC, "can't alloc memory\n");
	}
	SpaceAvailable = BufferSize;
    }

    ReturnValue = buffer;

    while (*s)
	*buffer++ = *s++;
    *buffer++ = 0;
    SpaceAvailable -= (len + 1);
    return ReturnValue;
}

void 
__MrmPush(char *s)
{
    __MrmStack[__MrmStackPointer] = s;
    __MrmStackPointer++;
    if (__MrmStackPointer >= MAX_STACK)
    {
	__MrmExit(LOC, "STACK OVERFLOW");
    }
}

char *
__MrmPop(void)
{
    __MrmStackPointer--;
    if (__MrmStackPointer < 0)
    {
	__MrmExit(LOC, "STACK UNDERFLOW\n");
    }
    return __MrmStack[__MrmStackPointer];
}

void 
__MrmExit(int line, char *file, const char *fmt,...)
{
    va_list ap;

    va_start(ap, fmt);
    fprintf(stderr, "%s:%d:", file, line);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
    abort();			/* Get a dump so we can see the stack frame, variables, etc... */
}

void 
__MrmWarn(int line, char *file, const char *fmt,...)
{
    va_list ap;

    va_start(ap, fmt);
    fprintf(stderr, "%s:%d:", file, line);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
    return;
}

File: QMOLEDEV64/nmap-4.76/zenmap/radialnet/util/misc.py
# vim: set fileencoding=utf-8 :

# Copyright (C) 2007, 2008 Insecure.Com LLC.
#
# Author: João Paulo de Souza Medeiros <ignotus21@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

from radialnet.core.Coordinate import CartesianCoordinate
from radialnet.gui.RadialNet import *
from radialnet.util.geometry import *
import math



def ipv4_compare(ip1, ip2):
    """
    """
    ip1 = [int(i) for i in ip1.split('.')]
    ip2 = [int(i) for i in ip2.split('.')]

    for i in range(4):

        if ip1[i] != ip2[i]:

            if ip1[i] < ip2[i]:
                return -1

            else:
                return 1

    return 0


def list_difference_update(list, difference):
    """
    """
    for item in difference:

        if item in list:
            list.remove(item)


def list_update(list, append):
    """
    """
    for item in append:

        if item not in list:
            list.append(item)


def swap(list, a, b):
    """
    """
    list[a], list[b] = list[b], list[a]


def sort_children(children, father):
    """
    """
    if len(children) < 2:
        return children

    # create angle reference
    f_x, f_y = father.get_cartesian_coordinate()

    for child in children:

        c_x, c_y = child.get_cartesian_coordinate()
        _, angle = CartesianCoordinate(c_x - f_x, c_y - f_y).to_polar()

        child.set_draw_info({'angle_from_father': math.degrees(angle)})

    return sort_children_by_angle(children)


def sort_children_by_angle(children):
    """
    """
    if len(children) < 2:
        return children

    vector = list()
    vector.append(children.pop())

    for a in children:

        theta_a = normalize_angle(a.get_draw_info('angle_from_father'))

        for i in range(len(vector) -1, -1, -1):

            b = vector[i]

            theta_b = normalize_angle(b.get_draw_info('angle_from_father'))

            if theta_b <= theta_a <= theta_b + 180:

                vector.insert(i + 1, a)
                break

        else:
            vector.insert(0, a)

    return vector
